src/v/compression/async_stream_zstd.cc

```diff
@@ -0,0 +1,201 @@
+
+    ss::temporary_buffer<char> d_buffer;
+    ss::temporary_buffer<char> c_buffer;
+};
+
+static thread_local zstd_workspace workspace;
+static thread_local bool workspace_initalized = false;
+
+static constexpr size_t default_decompression_size = 2_MiB;
+static constexpr int default_compression_level = 3;
+
+void async_stream_zstd::init_workspace(size_t decompression_size) {
+    if (workspace_initalized) {
+        return;
+    }
+
+    workspace.d_buffer = ss::temporary_buffer<char>(ZSTD_DStreamOutSize());
+    workspace.c_buffer = ss::temporary_buffer<char>(ZSTD_CStreamOutSize());
+
+    auto d_workspace_size = ZSTD_estimateDStreamSize(decompression_size);
+    auto c_workspace_size = ZSTD_estimateCStreamSize(default_compression_level);
+
+    // zstd requires alignment on both of the buffer below.
+    workspace.d_workspace = ss::allocate_aligned_buffer<char>(
+      d_workspace_size, 8);
+    workspace.c_workspace = ss::allocate_aligned_buffer<char>(
+      c_workspace_size, 8);
+
+    vassert(
+      workspace.c_workspace,
+      "Failed to allocate zstd workspace of size {}",
+      c_workspace_size);
+    vassert(
+      workspace.d_workspace,
+      "Failed to allocate zstd workspace of size {}",
+      d_workspace_size);
+
+    workspace.compress_ctx = ZSTD_initStaticCCtx(
+      workspace.c_workspace.get(), c_workspace_size);
+    workspace.decompress_ctx = ZSTD_initStaticDCtx(
+      workspace.d_workspace.get(), d_workspace_size);
+
+    throw_if_error(ZSTD_CCtx_setParameter(
+      workspace.compress_ctx,
+      ZSTD_c_compressionLevel,
+      default_compression_level));
+
+    workspace_initalized = true;
+}
+
+ss::future<iobuf> async_stream_zstd::do_compress(const iobuf& x) {
+    init_workspace(default_decompression_size);
+
+    auto lock = co_await workspace.compress_mtx.get_units();
+
+    // reset compressor ctx
+    ZSTD_CCtx_reset(workspace.compress_ctx, ZSTD_reset_session_only);
+
+    // NOTE: always enable content size. **decompression** depends on this
+    throw_if_error(
+      ZSTD_CCtx_setPledgedSrcSize(workspace.compress_ctx, x.size_bytes()));
+
+    iobuf ret;
+
+    for (auto it = x.cbegin(); it != x.cend();) {
+        ZSTD_inBuffer in = {.src = it->get(), .size = it->size(), .pos = 0};
+        ++it;
+
+        const auto mode = (it == x.cend()) ? ZSTD_e_end : ZSTD_e_continue;
+
+        int finished;
+        do {
+            ZSTD_outBuffer out = {
+              .dst = workspace.c_buffer.get_write(),
+              .size = workspace.c_buffer.size(),
+              .pos = 0};
+
+            const auto remaining = ZSTD_compressStream2(
+              workspace.compress_ctx, &out, &in, mode);
+            throw_if_error(remaining);
+
+            ret.append(workspace.c_buffer.get(), out.pos);
+            finished = (mode == ZSTD_e_end) ? (remaining == 0)
+                                            : (in.pos == in.size);
+
+            co_await ss::coroutine::maybe_yield();
```

Comment:
We take the input buffer by reference. This seems like a problem if we yield, right? Since now the caller will have to organize it such that the iobuf lives until the compression is complete, which may be not be straightfoward.

`iobuf` already supports cheap move semantics, right? Maybe we should have a version where you move in the iobuf then you don't need to worry about the lifetime of the input.

Reply:
Switched over to taking by value in a co-routine. Thanks!